packages/coinjoin/src/client/CoinjoinRound.ts

```diff
@@ -198,18 +186,78 @@ export class CoinjoinRound extends EventEmitter implements SerializedCoinjoinRou
+        if (this.phase === RoundPhase.InputRegistration) {
+            return inputRegistration(this, prison, processOptions);
+        }
+        if (this.phase === RoundPhase.ConnectionConfirmation) {
+            return connectionConfirmation(this, processOptions);
+        }
+        if (this.phase === RoundPhase.OutputRegistration) {
+            return outputRegistration(this, accounts, prison, processOptions);
+        }
+        if (this.phase === RoundPhase.TransactionSigning) {
+            return transactionSigning(this, processOptions);
+        }
+        if (this.phase === RoundPhase.Ended) {
+            return Promise.resolve(this);
+        }
+        return Promise.resolve(this);
     }
 
     getRequest(): CoinjoinRequestEvent | void {
-        // TODO
+        if (this.phase === RoundPhase.InputRegistration) {
+            const inputs = this.inputs.filter(input => !input.ownershipProof && !input.requested);
+            if (inputs.length > 0) {
+                inputs.forEach(input => {
+                    this.options.log(`Requesting ownership for <${input.outpoint}>`);
+                    input.setRequest('ownership');
+                });
+                return {
+                    type: 'ownership',
+                    roundId: this.id,
+                    inputs,
+                    commitmentData: this.commitmentData,
+                };
+            }
+        }
+        if (this.phase === RoundPhase.TransactionSigning) {
+            const inputs = this.inputs.filter(i => !i.witness && !i.requested);
+            if (inputs.length > 0 && this.transactionData) {
+                inputs.forEach(input => {
+                    this.options.log(`Requesting witness for <${input.outpoint}>`);
+                    input.setRequest('signature');
+                });
+                return {
+                    type: 'signature',
+                    roundId: this.id,
+                    inputs,
+                    transaction: this.transactionData,
+                };
+            }
+        }
     }
 
-    resolveRequest(_: CoinjoinResponseEvent) {
-        // TODO
+    resolveRequest({ type, inputs }: CoinjoinResponseEvent) {
+        const { log } = this.options;
+        inputs.forEach(i => {
+            const input = this.inputs.find(a => a.outpoint === i.outpoint);
+            // reset request in input
+            input?.setRequest();
+            if ('error' in i) {
+                log(`Resolving ${type} request for <${i.outpoint}> with error. ${i.error}`);
+                input?.setError(new Error(i.error));
+            } else if ('ownershipProof' in i) {
+                log(`Resolving ${type} request for <${i.outpoint}>`);
+                input?.setOwnershipProof(i.ownershipProof);
+            } else if ('signature' in i) {
+                log(`Resolving ${type} request for <${i.outpoint}>`);
+                input?.setWitness(i.signature, i.index);
+            }
+        });
```

Comment:
nitpick: would add something like `  if (!input) return;` and get rid of optional chaining operators

```suggestion
     inputs.forEach(i => {
            const input = this.inputs.find(a => a.outpoint === i.outpoint);

            if (!input) return;

            // reset request in input
            input.setRequest();
            if ('error' in i) {
                log(`Resolving ${type} request for <${i.outpoint}> with error. ${i.error}`);
                input.setError(new Error(i.error));
            } else if ('ownershipProof' in i) {
                log(`Resolving ${type} request for <${i.outpoint}>`);
                input.setOwnershipProof(i.ownershipProof);
            } else if ('signature' in i) {
                log(`Resolving ${type} request for <${i.outpoint}>`);
                input.setWitness(i.signature, i.index);
            }
        });
```